site/en/docs/extensions/mv3/tut_puppeteer-testing/index.md

```diff
@@ -0,0 +1,177 @@
+following:
+
+```js
+{
+  "name": "puppeteer-demo",
+  "version": "1.0"
+}
+```
+
+Similar to an extension's manifest.json file, this file is required by all Node projects.
+
+### Step 2: Install Puppeteer and Jest
+
+Run `npm install puppeteer jest` to add Puppeteer and Jest as dependencies. They will be
+automatically added to your `package.json` file.
+
+It is possible to write standalone Puppeteer tests, but we'll use Jest as a test runner to provide
+some additional structure to our code.
+
+### Step 3: Create an entry point
+
+In a new file called `index.test.js`, add the following:
+
+{% Aside %}
+Jest adds `beforeEach` and `afterEach` to our environment when running the tests, so we do not need
+to import them. However, you may need to configure [ESLint][eslint] or [TypeScript][typescript]
+accordingly.
+{% endAside %}
+
+```js
+const puppeteer = require('puppeteer');
+
+const EXTENSION_PATH = '../../api-samples/history/showHistory';
+const EXTENSION_ID = 'jkomgjfbbjocikdmilgaehbfpllalmia';
+
+let browser;
+
+beforeEach(async () => {
+  // TODO: Launch the browser.
+});
+
+afterEach(async () => {
+  // TODO: Close the browser.
+});
+```
+
+### Step 4: Launch the browser
+
+{% Aside %}
+In this example, `headless` is set to `false`. This causes the browser window to be visible while
+the tests are running which can be helpful during development. Outside of development, consider
+setting it to `'new'` which uses Chrome's [new headless mode][new-headless].
+{% endAside %}
+
+Update `beforeEach` and `afterEach` to launch and close the browser. When running many tests, you
+may wish to consider using the same browser. However, this is generally discouraged as it reduces
+the isolation between your tests and may cause one test to impact the outcome of another.
+
+```js
+beforeEach(async () => {
+  browser = await puppeteer.launch({
+    headless: false,
+    args: [
+      `--disable-extensions-except=${EXTENSION_PATH}`,
+      `--load-extension=${EXTENSION_PATH}`
+    ]
+  });
+});
+
+afterEach(async () => {
+  await browser.close();
+  browser = undefined;
+});
+```
+
+### Step 5: Add an alias
+
+To make running the tests easier, add an alias to your `package.json` file:
+
+```js
+{
+  "name": "puppeteer-demo",
+  "version": "1.0",
+  "dependencies": {
+    "puppeteer": "^21.3.6"
+  },
+  "scripts": {
+    "start": "jest ."
+  }
+}
+```
+
+This will run all files ending in `.test.js` in the current directory.
+
+### Step 6: Open the popup
+
+Let's add a basic test that opens the popup in a new page. We need to do this because Puppeteer
+does not support accessing an extension popup from the popup window. Add the following code:
+
+```js
+test('popup renders correctly', async () => {
+  const page = await browser.newPage();
+  await page.goto(`chrome-extension://${EXTENSION_ID}/popup.html`);
+});
+```
+
+The method `browser.newPage()` actually supports an optional URL parameter, but unfortunately this
```

Comment:
Optional: add a link to `browser.newPage()`?

Reply:
Realised this paragraph was wrong so dropped it entirely.